Issue #5277: Sorted Aggregate Sorting  …

hawkfish opened the pull request:
Convert to the new sorting logic.
Remove dead code (ChunkCollection::Sort).
Disable busted test.

Commit:
Issue #5277: Sorted Aggregate Sorting

Convert to the new sorting logic.

Commit:
Issue #5277: Sorted Aggregate Sorting

Remove dead code (ChunkCollection::Sort)

Mytherin committed:
Correctly accept BUILD_JEMALLOC_EXTENSION on Linux

Mytherin committed:
Have to move this up

Commit:
Merge branch 'master' into orderagg-sorter

Commit:
Issue #5277: Sorted Aggregate Sorting

Disable busted test.

hawkfish commented:
fixes #5347

jfb-h committed:
Fix docstring of `load!` and relax types

Calling `load!` fails due to the type restriction on `con`, probably due to confusion about `DBInterface.Connection` vs `DuckDB.Connection`. The easiest fix is to just omit the type restriction.

Mytherin committed:
Merge pull request #5343 from Mytherin/jemallocfix

Correctly accept BUILD_JEMALLOC_EXTENSION on Linux

Mytherin commented:
Thanks for the PR! Could we remove the entire `ChunkCollection` class after this?

I also wonder about performance - have you run any tests to see how this compares to the previous implementation for very small data sets (e.g. 1-3 rows per group)?

jeremiahpslewis committed:
Bump DuckDB_jll compat

lnkuiper commented:
We could consider adding `DataChunk::Sort` that does zero-copy (slicing instead) sorting for small sorts. Would be happy to implement that

jeremiahpslewis committed:
Update Project.toml

Mytherin committed:
Merge pull request #5354 from jfb-h/patch-2

[julia] fix docstring of `load!` and relax type restriction

Mytherin committed:
Merge pull request #5356 from jeremiahpslewis/patch-2

Bump DuckDB_jll compat to v0.6

Tishj committed:
fixed Appender for decimal columns

hawkfish commented:
> We could consider adding `DataChunk::Sort` that does zero-copy (slicing instead) sorting for small sorts. Would be happy to implement that

Amusingly that's [how I implemented sorting](https://github.com/hawkfish/olapscript/blob/main/src/table.js#L979) in OLAPScript! So please go ahead. I just want to make sure it handles all the sorting constraints (part of what I removed was a common hack for `ChunkCollection::Sort` to handle descending and null placement).

For me the questions are: 

* How many different sort algorithms should we have?
* If we have more than one, where is the cut off?

@Mytherin asked about getting rid of `ChunkCollection` entirely, and we could get rid of it here by dumping values into a `DataChunk` and then switching over to `LocalSortState` when it got "too big" (especially since we can now use arbitrary sized chunks). The only other place where `ChunkCollection` is used right now is in reservoir sampling, but that seems out of scope (but the next target.)

Mytherin commented:
> I'd use the existing sort code in pretty much the same way as list_sort does, would that count as 1 or 2 implementations?

I was trying to benchmark to see how this compared to the current ChunkCollection code and actually found a bug in  `list_sort` - the following code crashes:

```sql
create table tbl as select i from range(1000000) t(i);
select list_sort([i]) from tbl;
```

Strangely enough this works if the number of rows is changed to 100K.

lnkuiper commented:
I ran this in debug mode and it seems like the `LocalSortState` is destroyed while data is sunk in there.

Tishj committed:
clear up ambiguity issue that's making Linux compilation fail

Commit:
Merge branch 'orderagg-sorter' of github.com:hawkfish/duckdb into orderagg-sorter

hawkfish commented:
> I'd use the existing sort code in pretty much the same way as list_sort does, would that count as 1 or 2 implementations?

Just one. I meant actual sorting APIs (Laurens', `ChunkCollection`).

Commit:
Issue #5277: Sorted Aggregate Sorting

Replace ChunkCollection with ColumnDataCollection.
Buffer small argument counts to avoid CDC allocation for small groups.

Commit:
Issue #5342: DATE_PART Struct Indexing

Fix the result indexing for the Struct variants of DATE_PART.
The existing code wrote to the same output row as the input selection,
which was not correct.

Tishj committed:
remove using namespace std, conflicting names 'align' - and also just bad practice

Tishj committed:
include windows.hpp, to avoid collision with 'max' defined by windows if NOMINMAX is not set - though I doubt this fixes it if windows.h is already included by the user before including duckdb

Mytherin committed:
Merge pull request #5382 from hawkfish/sparse-datepart

Issue #5342: DATE_PART Struct Indexing

ak-coram committed:
Add reference to cleanup function for duckdb_result_get_chunk

Tishj committed:
create distinction between logical and physical appenders

Tishj committed:
make date_t::infinity/ninfinity non-inlined

Tishj committed:
make some LONG tests into '.test_slow' while I'm at it

Tishj committed:
rename 'align' -> 'align_value'

Tishj committed:
add a test that includes windows.h before duckdb.hpp

Tishj committed:
give timestamp_t the same treatment

Mytherin committed:
Fix #5371: correctly use instance cache

Mytherin committed:
Merge pull request #5389 from ak-coram/api-docs-get-chunk-cleanup

Add reference to cleanup function for duckdb_result_get_chunk

Mytherin committed:
Fix #5390: in filter pull-up optimizer avoid adding columns to one side of a set operation

Mytherin committed:
Add test

Commit:
Merge branch 'master' into orderagg-sorter

Commit:
Issue #5277: Sorted Aggregate Sorting

Fix DataChunk lazy initialisation.

Mytherin committed:
Merge pull request #5400 from Mytherin/issue5390

Fix #5390: in filter pull-up optimizer avoid adding columns to one side of a set operation

Mytherin committed:
Merge pull request #5398 from Mytherin/issue5371

Fix #5371: correctly use instance cache in JDBC and ODBC connector

Mause committed:
fix: upgrade npm's internal node-gyp

many thanks to @archiewood from Evidence.dev who found the fix for this

lnkuiper commented on src/function/aggregate/sorted_aggregate_function.cpp:
```diff
@@ -58,26 +46,102 @@ struct SortedAggregateBindData : public FunctionData {
+			const auto chunk_count = ordering->ChunkCount();
+			for (idx_t chunk_idx = 0; chunk_idx < chunk_count; ++chunk_idx) {
+				ordering->FetchChunk(chunk_idx, sort_buffer);
+				arguments->FetchChunk(chunk_idx, arg_buffer);
```
`FetchChunk` is cannot do zero-copy reads from the CDC, and is therefore much less efficient than scanning the CDC using `ColumnDataCollection::Scan(ColumnDataScanState &state, DataChunk &result)`. Doing the zero-copy read here is fine since the data is copied to the sort state when you sink it anyway.

hawkfish replied:
Thanks - this is why I dragged you in!

hawkfish replied:
Ah when I do that the sort code crashes with a bad memory access.

hawkfish replied:
My bad - I wasn't Resetting the DataChunks.

lnkuiper reviewed (commented):
Looks good! I just have one comment

Tishj committed:
moved date_t and timestamp_t into date.hpp and timestamp.hpp respectively

Mytherin committed:
Add support for reading JSON from Parquet files

Tishj committed:
moved some stuff around so things actually compile

Tishj committed:
missed a couple spots

Tishj committed:
add appender tests for decimals to the C api

Mytherin committed:
Merge pull request #5401 from Mytherin/jsonconvertedtype

Add support for reading JSON type columns from Parquet files

Commit:
Merge branch 'master' into orderagg-sorter

Commit:
Merge branch 'orderagg-sorter' of https://github.com/hawkfish/duckdb into orderagg-sorter

Tishj committed:
forgot to destroy the appender

Commit:
Issue #5277: Sorted Aggregate Slicing

Slice in place.

Mytherin committed:
Merge pull request #5386 from Tishj/test_buffer_manager_no_using_std

[Dev] Fix compilation issues related to MSVC and Windows.h

Mytherin committed:
Merge pull request #5402 from Mause/patch-8

fix: upgrade npm's internal node-gyp

Commit:
Issue #5277: Sorted Aggregate Scans

Use Scans correctly (with Reset()).

Commit:
fix bug with loss of order preservation in insert

Mytherin committed:
Merge pull request #5364 from Tishj/appender_into_decimal

[Appender] Appender can now properly append to DECIMAL columns

Mytherin committed:
In Allocator::Allocate, through a std::bad_alloc exception if allocation fails

Mytherin committed:
Also throw in ReallocateData

Mytherin committed:
Throw internal exceptions on giant allocations (as they are likely the result of an unsigned integer underflow)

Mytherin committed:
Merge pull request #5427 from lnkuiper/parallel_unionall_insert

Fix bug causing loss of order preservation in insert

Mytherin committed:
Throw internal exception on nullptr in AllocatedData

Mytherin committed:
Remove premature optimization - exceptions are always unlikely

Commit:
Merge branch 'master' into orderagg-sorter

Commit:
Merge branch 'orderagg-sorter' of github.com:hawkfish/duckdb into orderagg-sorter

Commit:
Issue #5277: Sorted Aggregate Benchmark

Add a new benchmark for large numbers of groups.
Performance seems unchanged from the old sort code.

Mytherin committed:
Merge pull request #5439 from Mytherin/badalloc

Allocator: throw std::bad_alloc if a malloc allocation fails

lokax committed:
Forbit ORDER BY Columns(...)

Mytherin committed:
Merge pull request #5444 from lokax/fix-order-by-columns-star

Fix the use of COLUMNS(...) in ORDER BY clause

Commit:
Merge branch 'master' into orderagg-sorter

Commit:
Merge branch 'orderagg-sorter' of https://github.com/hawkfish/duckdb into orderagg-sorter
